ring/replication_set.go

```diff
@@ -140,11 +154,27 @@ func DoUntilQuorum[T any](ctx context.Context, r ReplicationSet, f func(context.
 		contextTracker = newDefaultContextTracker(ctx, r.Instances)
 	}
 
+	if minimizeRequests {
+		resultTracker.releaseMinimumRequests()
+	} else {
+		resultTracker.releaseAllRequests()
+	}
+
 	for i := range r.Instances {
 		instance := &r.Instances[i]
 		instanceCtx := contextTracker.contextFor(instance)
 
 		go func(desc *InstanceDesc) {
+			if !resultTracker.awaitRelease(desc) {
```

Comment:
Below, when the context is canceled we call `cleanupResultsAlreadyReceived()` and return. I've the feeling we may have a condition where this goroutine is never terminated because `awaitRelease()` waits indefinitely in that case.

[non blocking] Generally speaking, I've found quite hard to follow the "failure" case to make sure `awaitRelease()` always return in every case. I'm wondering if there's anything we can do to make the code easier to follow.

Reply:
> Below, when the context is canceled we call `cleanupResultsAlreadyReceived()` and return. I've the feeling we may have a condition where this goroutine is never terminated because `awaitRelease()` waits indefinitely in that case.

Good catch, yep, that could block forever. I'll add a test for this case and then fix it.

> [non blocking] Generally speaking, I've found quite hard to follow the "failure" case to make sure `awaitRelease()` always return in every case. I'm wondering if there's anything we can do to make the code easier to follow.

We might be able to solve both of these problems in one go: I'll modify `awaitRelease()` to take the request context as well, so if that is cancelled, it'll return immediately. This should make it easier to confirm that `awaitRelease()` will always return.

Reply:
I've implemented this (f272c76), however it introduces additional complexity for the tests (93656f2):

Because `awaitStart()` (formerly `awaitRelease()`) now returns if the instance's context has been cancelled, there's a chance that `f` is not always called and therefore won't need to be cleaned up. The tests previously assumed this was more deterministic, and so the assertions around cleanup behaviour are now more complex.